File: Course_2-Data_Structures/Week-4/Excercise_Challenges/3_hash_substring/hash_substring.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int hash_func(string s)
{
    int hash = 0;
    for(int i = 0; i < s.length(); ++i)
        hash += int(s[i]);

    return hash;
}

vector<int> precompute_hashes(string p, string t)
{
    vector<int> h;
    int pl = p.length(), bound = t.length() - p.length() + 1;

    h.push_back(hash_func(t.substr(0, pl)));
    
    for(int i = 1; i < bound; ++i){
        int old_hash = h[i - 1];
        int left_hash = int(t[i - 1]);
        int right_hash = int(t[i + pl - 1]);

        int ith_hash = old_hash - left_hash + right_hash;

        h.push_back(ith_hash);
    }

    return h;
}

vector<int> get_occurrences_rabin_karp(string p, string t)
{
    vector<int> ans;
    int pattern_hash = hash_func(p);

    vector<int> H = precompute_hashes(p, t);

    for(int i = 0; i <= t.length() - p.length(); ++i){
        if(pattern_hash == H[i]){
            string s = t.substr(i, p.length());
            if(p == s)
                ans.push_back(i);
        }
    }

    return ans;
}

vector<int> get_occurrences_naive(string p, string t)
{    
    vector<int> ans;
    for(int i = 0; i + p.size() <= t.size(); ++i)
        if(t.substr(i, p.size()) == p)
            ans.push_back(i);
    return ans;
}


int main(void)
{
    ios_base::sync_with_stdio(false);
    string pattern, text;
    cin >> pattern;
    cin >> text;

    vector<int> res = get_occurrences_rabin_karp(pattern, text);
    for(int i = 0; i < res.size(); ++i)
        cout << res[i] << " ";
    cout << endl;

    return 0;
}

File: Course_2-Data_Structures/Week-4/Excercise_Challenges/3_hash_substring/hash_substring.py
class RabinKarp:
    def __init__(self):
        self.pattern = ''
        self.text = ''
        self.pattern_len = 0
        self.text_len = 0
        self.bound = 0
        self.hash_list = []
        self.find_index = []

    def hash_func(self, s):
        h = 0
        for a in s:
            h += ord(a)

        return h

    def precompute_hashes(self):
        self.hash_list = [self.hash_func(self.text[:self.pattern_len])]
        
        for i in range(1, self.bound):
            old_hash = self.hash_list[i - 1]
            left_hash = ord(self.text[i - 1])
            right_hash = ord(self.text[i + self.pattern_len - 1])

            total_hash = old_hash - left_hash + right_hash
            self.hash_list.append(total_hash)

    def get_occurrences(self):
        pattern_hash = self.hash_func(self.pattern)

        self.precompute_hashes()
        for i in range(self.bound):
            if pattern_hash == self.hash_list[i]:
                if self.pattern == self.text[i : i + self.pattern_len]:
                    self.find_index.append(i)

    def read_input(self):
        self.pattern = input()
        self.text = input()

        self.pattern_len = len(self.pattern)
        self.text_len = len(self.text)
        self.bound = self.text_len - self.pattern_len + 1

    def print_occurrences(self):
        print(" ".join(map(str, self.find_index)))

    def solve(self):
        self.read_input()
        self.get_occurrences()
        self.print_occurrences()


if __name__ == '__main__':
    obj = RabinKarp()
    obj.solve()


